engine/table/src/main/java/io/deephaven/engine/table/impl/remote/ConstructSnapshot.java

```diff
@@ -1509,61 +1531,184 @@ private static boolean serializeAllTable(
-
-                final BarrageMessage.ModColumnData mcd = new BarrageMessage.ModColumnData();
-                snapshot.modColumnData[ii] = mcd;
-                mcd.rowsModified = RowSetFactory.empty();
-                mcd.data = getSnapshotDataAsChunkList(sourceToUse, null, RowSetFactory.empty(), usePrev);
-                mcd.type = acd.type;
-                mcd.componentType = acd.componentType;
-                mcd.chunkType = sourceToUse.getChunkType();
+            } else {
+                // Snapshot all non-empty columns serially
+                snapshotColumnsSerial(columnSources, nonEmptyColumnsIndices, table, usePrev, snapshot);
             }
         }
-
         if (log.isDebugEnabled()) {
             final LogEntry logEntry = log.debug().append(System.identityHashCode(logIdentityObject))
                     .append(": Snapshot candidate step=")
                     .append((usePrev ? -1 : 0) + LogicalClock.getStep(getConcurrentAttemptClockValue()))
                     .append(", rows=").append(snapshot.rowsIncluded).append("/").append(keysToSnapshot)
                     .append(", cols=");
-            if (columnsToSerialize == null) {
+            if (columnsToSnapshot == null) {
                 logEntry.append("ALL");
             } else {
-                logEntry.append(FormatBitSet.formatBitSet(columnsToSerialize));
+                logEntry.append(FormatBitSet.formatBitSet(columnsToSnapshot));
             }
             logEntry.append(", usePrev=").append(usePrev).endl();
         }
+        return true;
+    }
 
+    /**
+     * Check if all the required column sources are in memory and should allow efficient access.
+     */
+    private static boolean allColumnSourcesInMemory(
+            @NotNull final Table table,
+            @NotNull final String[] columnSources,
+            @NotNull final TIntList columnIndicesToSnapshot) {
+        final int numColumnsToSnapshot = columnIndicesToSnapshot.size();
+        for (int colRank = 0; colRank < numColumnsToSnapshot; ++colRank) {
+            final ColumnSource<?> columnSource =
+                    table.getColumnSource(columnSources[columnIndicesToSnapshot.get(colRank)]);
+            if (!isColumnSourceInMemory(columnSource)) {
+                return false;
+            }
+        }
         return true;
     }
 
-    private static boolean serializeAllTables(
+    /**
+     * Recursively check if the column source is in-memory or redirected to an in-memory column source.
+     */
+    private static boolean isColumnSourceInMemory(final ColumnSource<?> columnSource) {
+        if (columnSource instanceof InMemoryColumnSource) {
+            return true;
+        }
+        if (!(columnSource instanceof RedirectedColumnSource)) {
+            return false;
+        }
+        final ColumnSource<?> innerSource = ((RedirectedColumnSource<?>) columnSource).getInnerSource();
+        return isColumnSourceInMemory(innerSource);
+    }
```

Comment:
While performance is terrible with many redirected column sources wrapping each other, we should avoid recursion whenever it is easy to do so.

Also, note that `InMemoryColumnSource` has a `#isInMemory` call you also need to check.

```suggestion
    private static boolean isColumnSourceInMemory(ColumnSource<?> columnSource) {
        do {
            if (columnSource instanceof InMemoryColumnSource) {
                return ((InMemoryColumnSource) columnSource).isInMemory();
            }
            if (!(columnSource instanceof RedirectedColumnSource)) {
                return false;
            }
            columnSource = ((RedirectedColumnSource<?>) columnSource).getInnerSource();
        } while (true);
    }
```